libethereum/WarpPeerCapability.cpp

```diff
@@ -0,0 +1,138 @@
+/*
+	This file is part of cpp-ethereum.
+
+	cpp-ethereum is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	cpp-ethereum is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with cpp-ethereum.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "WarpPeerCapability.h"
+#include "SnapshotStorage.h"
+
+#include "libethcore/Exceptions.h"
+
+namespace dev
+{
+namespace eth
+{
+
+WarpPeerCapability::WarpPeerCapability(std::shared_ptr<p2p::SessionFace> _s, p2p::HostCapabilityFace* _h, unsigned _i, p2p::CapDesc const& /* _cap */):
+	Capability(_s, _h, _i)
+{
+}
+
+void WarpPeerCapability::init(unsigned _hostProtocolVersion, u256 _hostNetworkId, u256 _chainTotalDifficulty, h256 _chainCurrentHash, h256 _chainGenesisHash, std::shared_ptr<SnapshotStorageFace const> _snapshot)
+{
+	m_snapshot = _snapshot;
+	
+	h256 snapshotBlockHash;
+	u256 snapshotBlockNumber;
+	if (m_snapshot)
+	{
+		bytes const snapshotManifest(m_snapshot->readManifest());
+		RLP manifest(snapshotManifest);
+		if (manifest.itemCount() != 6)
+			BOOST_THROW_EXCEPTION(InvalidSnapshotManifest());
+		snapshotBlockNumber = manifest[4].toInt<u256>(RLP::VeryStrict);
+		snapshotBlockHash = manifest[5].toHash<h256>(RLP::VeryStrict);
+	}
+
+	requestStatus(_hostProtocolVersion, _hostNetworkId, _chainTotalDifficulty, _chainCurrentHash, _chainGenesisHash, snapshotBlockHash, snapshotBlockNumber);
+}
+
+bool WarpPeerCapability::interpret(unsigned _id, RLP const& _r)
+{
+	if (!m_snapshot)
+		return false;
+
+	try
+	{
+		switch (_id)
+		{
+		case WarpStatusPacket:
+		{
+			if (_r.itemCount() < 7)
+				BOOST_THROW_EXCEPTION(InvalidWarpStatusPacket());
+
+			auto const protocolVersion = _r[0].toInt<unsigned>(RLP::VeryStrict);
+			auto const networkId = _r[1].toInt<u256>(RLP::VeryStrict);
+			auto const totalDifficulty = _r[2].toInt<u256>(RLP::VeryStrict);
+			auto const latestHash = _r[3].toHash<h256>(RLP::VeryStrict);
+			auto const genesisHash = _r[4].toHash<h256>(RLP::VeryStrict);
```

Comment:
Is it a good idea to check `genesisHash` against `_chainGenesisHash` given to `init()`?  (Likewise for `networkId`?)

Reply:
Yes, I do that in the snapshot downloading PRs.

In this PR it doesn't make much difference, because it's a server part, and downloading peer checks it anyway.
I will add it later, anyway

Reply:
OK.